### PR TITLE
Integration for TensorFlow's Estimator API

### PR DESCRIPTION
**Pull request by sfujiwara**
I implemented `TensorFlowPruningHook` for TensorFlow's Estimator API with [SessionRunHook](https://www.tensorflow.org/api_docs/python/tf/train/SessionRunHook).
We can use it as below (full example is [here](https://gist.github.com/sfujiwara/fa3b763943a1d4758a01c89a8f3df2a5)):

```python
def objective(trial):
    save_steps = 50
    # Create input functions for train and eval
    train_input_fn, eval_input_fn = create_input_fn()
    # Hyper parameters to be tuned with Optuna
    learning_rate = trial.suggest_loguniform("learning_rate", 1e-5, 1e-2)
    # Create Estimator config
    config = tf.estimator.RunConfig(save_summary_steps=save_steps, save_checkpoints_steps=save_steps)
    # Create Estimator
    clf = tf.estimator.DNNClassifier(
        feature_columns=[tf.feature_column.numeric_column(key="x", shape=[4])],
        n_classes=3,
        hidden_units=[],
        optimizer=tf.train.GradientDescentOptimizer(learning_rate=learning_rate),
        model_dir="outputs_pruning/lr_{}".format(learning_rate),
        config=config
    )
    # Create hooks
    early_stopping_hook = tf.contrib.estimator.stop_if_no_decrease_hook(clf, "accuracy", save_steps)
    optuna_pruning_hook = TensorFlowPruningHook(
        trial=trial,
        estimator=clf,
        metrics_name="accuracy",
        is_higher_better=True,
        run_every_steps=10,
    )
    hooks = [early_stopping_hook, optuna_pruning_hook]
    # Create TrainSpec and EvalSpec
    train_spec = tf.estimator.TrainSpec(input_fn=train_input_fn, max_steps=500, hooks=hooks)
    eval_spec = tf.estimator.EvalSpec(input_fn=eval_input_fn, steps=1000, start_delay_secs=0, throttle_secs=0)
    # Run training and evaluation
    tf.estimator.train_and_evaluate(clf, train_spec, eval_spec)
    result = clf.evaluate(input_fn=eval_input_fn, steps=100)
    accuracy = result["accuracy"]
    return 1.0 - accuracy
```

Note that multi-node distributed learning is not considered yet.

### CODE REVIEW

File: optuna/integration/tensorflow.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import optuna
+
+
+try:
+    import tensorflow as tf
+    from tensorflow.train import SessionRunHook
+    _available = True
+except ImportError as e:
+    _import_error = e
+    # PruningHook is disabled because TensorFlow is not available.
```
Comment (toshihikoyanase) [inline optuna/integration/tensorflow.py]: I think this `PruningHook` means `TensorFlowPruningHook`. Could you use the full name of hook as`TensorFlowPruningHook` like other integration modules?

File: optuna/integration/tensorflow.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    _available = False
+    SessionRunHook = object
+
+
+class TensorFlowPruningHook(SessionRunHook):
+    """TensorFlow SessionRunHook to prune umpromising trials.
```
Comment (toshihikoyanase) [inline optuna/integration/tensorflow.py]: 'umpromising' seems a typo for 'unpromising'.

File: optuna/integration/tensorflow.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+    Example:
+
+        Add a pruning SessionRunHook for a TensorFlow's Estimator.
+
+        .. code::
+
+                optuna_pruning_hook = OptunaPruningHook(
```
Comment (toshihikoyanase) [inline optuna/integration/tensorflow.py]: `OptunaPruningHook` seems an old name of `TensorFlowPruningHook`.

File: optuna/integration/tensorflow.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                    clf,
+                    tf.estimator.TrainSpec(input_fn=train_input_fn, max_steps=500, hooks=hooks),
+                    eval_spec
+                )
+    Args:
+        trial:
+            A :class:`~optuna.trial.Trial` corresponding to the current evaluation of
+            the objective function.
+        estimator:
+            A :estimator which you will use.
```
Comment (toshihikoyanase) [inline optuna/integration/tensorflow.py]: I think 'A :estimator' is a typo for 'An estimator'.

File: optuna/integration/tensorflow.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        metric:
+            An evaluation metric for pruning, e.g., ``accuracy`` and ``loss``.
+        is_higher_better:
+           It should be True if you use a metric to be maximize such as ``loss``.
+        run_every_steps:
+           An interval to watch the summary file.
+    """
+    def __init__(self, trial, estimator, metric, is_higher_better, run_every_steps):
```
Comment (toshihikoyanase) [inline optuna/integration/tensorflow.py]: Could you add a type hint to `TensorFlowPruningHook.__init__`? To keep compatibility with Python 2.7, we use type comments as follows:
```
    def __init__(
            self,
            trial,  # type: optuna.trial.Trial
            ....
    ):
        # type: (...) -> None

        self.trial = trial
```
Please refer to [`Study.__init__`](https://github.com/pfnet/optuna/blob/master/optuna/study.py#L58) as an example.

File: optuna/integration/tensorflow.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        self.trial = trial
+        self.estimator = estimator
+        self.current_summary_step = -1
+        self.metric = metric
+        self.is_higher_better = is_higher_better
+        self._global_step_tensor = None
+        self._timer = tf.train.SecondOrStepTimer(every_secs=None, every_steps=run_every_steps)
+
+    def begin(self):
```
Comment (toshihikoyanase) [inline optuna/integration/tensorflow.py]: ditto.

File: optuna/integration/tensorflow.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        self._global_step_tensor = tf.train.get_global_step()
+
+    def before_run(self, run_context):
```
Comment (toshihikoyanase) [inline optuna/integration/tensorflow.py]: ditto.

File: optuna/integration/tensorflow.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        del run_context
+        return tf.train.SessionRunArgs(self._global_step_tensor)
+
+    def after_run(self, run_context, run_value):
```
Comment (toshihikoyanase) [inline optuna/integration/tensorflow.py]: ditto.

File: tests/integration_tests/test_tensorflow.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+except ImportError:
+    _available = False
+
+
+import optuna
+from optuna.integration import TensorFlowPruningHook
+from optuna.testing.integration import DeterministicPruner
+
+
+def fixed_value_input_fn():
```
Comment (toshihikoyanase) [inline tests/integration_tests/test_tensorflow.py]: Please add a type hint to `fixed_value_input_fn`.

File: optuna/integration/tensorflow.py (hunk 1 above)
Comment (sfujiwara) [reply optuna/integration/tensorflow.py]: I fixed it as you proposed.

File: optuna/integration/tensorflow.py (hunk 2 above)
Comment (sfujiwara) [reply optuna/integration/tensorflow.py]: I fixed the typo.

File: optuna/integration/tensorflow.py (hunk 3 above)
Comment (sfujiwara) [reply optuna/integration/tensorflow.py]: I fixed it.

File: optuna/integration/tensorflow.py (hunk 4 above)
Comment (sfujiwara) [reply optuna/integration/tensorflow.py]: I fixed it.

File: optuna/integration/tensorflow.py (hunk 5 above)
Comment (sfujiwara) [reply optuna/integration/tensorflow.py]: I added a type hint as below:

```python
# type: (optuna.trial.Trial, tf.estimator.Estimator, str, bool, int) -> None
```

File: optuna/integration/tensorflow.py (hunk 6 above)
Comment (sfujiwara) [reply optuna/integration/tensorflow.py]: I added a type hint as below:

```python
# type: () -> None
```

File: optuna/integration/tensorflow.py (hunk 7 above)
Comment (sfujiwara) [reply optuna/integration/tensorflow.py]: I added a type hint as below:

```python
# type: (tf.train.SessionRunContext) -> tf.train.SessionRunArgs
```

File: optuna/integration/tensorflow.py (hunk 8 above)
Comment (sfujiwara) [reply optuna/integration/tensorflow.py]: I added a type hint as below:

```
# type: (tf.train.SessionRunContext, tf.train.SessionRunValues) -> None
```

File: tests/integration_tests/test_tensorflow.py (hunk 9 above)
Comment (sfujiwara) [reply tests/integration_tests/test_tensorflow.py]: I added a type hint as below:

```python
# type: () -> typing.Tuple[typing.Dict[str, tf.Tensor], tf.Tensor]
```

**Comment by sfujiwara** [comment]
@toshihikoyanase Thank you for your review. I fixed the typos and added type hints. Please check them!

**Comment by toshihikoyanase** [comment]
Thank you for your swift action. Your change looks good to me. I'll try it with some examples.

By the way, Optuna's development members will take new year's holidays, so please give us some time for further reviews.

File: optuna/integration/tensorflow.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+        del run_context
+        return tf.train.SessionRunArgs(self._global_step_tensor)
+
+    def after_run(self, run_context, run_values):
+        # type: (tf.train.SessionRunContext, tf.train.SessionRunValues) -> None
+
+        global_step = run_values.results
+        # Get eval metrics every n steps
+        if self._timer.should_trigger_for_step(global_step):
```
Comment (toshihikoyanase) [inline optuna/integration/tensorflow.py]: last triggered steps were updated in `self._timer.should_trigger_for_step` automatically, but they weren't as far as I confirmed [the source code](https://github.com/tensorflow/tensorflow/blob/r1.12/tensorflow/python/training/basic_session_run_hooks.py#L118). 
Could you check this issue?

Comment (sfujiwara) [reply optuna/integration/tensorflow.py]: Thanks, you are right. I implemented it referring to the [code](https://github.com/tensorflow/tensorflow/blob/b9e6bbc95bcffa481d29e31b448a03a91ba17eac/tensorflow/contrib/estimator/python/estimator/early_stopping.py#L445-L446), but forgot to run `self._timer.update_last_triggered_step(global_step)`.
I will add a line, and commit it after passing the tests.

Comment (sfujiwara) [reply optuna/integration/tensorflow.py]: I fixed it!

File: optuna/integration/tensorflow.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+        # type: (tf.train.SessionRunContext) -> tf.train.SessionRunArgs
+
+        del run_context
+        return tf.train.SessionRunArgs(self._global_step_tensor)
+
+    def after_run(self, run_context, run_values):
+        # type: (tf.train.SessionRunContext, tf.train.SessionRunValues) -> None
+
+        global_step = run_values.results
+        # Get eval metrics every n steps
```
Comment (sile) [inline optuna/integration/tensorflow.py]: Please add trailing periods to here and [L81](https://github.com/pfnet/optuna/pull/292/files#diff-6777d6b27690da66a18f118b10e4a1dcR87), because it is an Optuna's coding convention.

**Comment by sile** [comment]
I think it is good to add `TensorFlowPruningHook ` to [the pruner integration list](https://optuna.readthedocs.io/en/stable/tutorial/pruning.html#integration-modules-for-pruning) by modifying `docs/source/tutorial/pruning.rst` file(cf. [A diff in Keras integration PR](https://github.com/pfnet/optuna/pull/256/files#diff-a1739713cf65447a882269f4f150efa8)).

File: optuna/integration/tensorflow.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    Args:
+        trial:
+            A :class:`~optuna.trial.Trial` corresponding to the current evaluation of
+            the objective function.
+        estimator:
+            An estimator which you will use.
+        metric:
+            An evaluation metric for pruning, e.g., ``accuracy`` and ``loss``.
+        is_higher_better:
+           It should be True if you use a metric to be maximize such as ``loss``.
```
[the remark this answers is not in the document]
Comment (sile) [reply optuna/integration/tensorflow.py]: Should it be minimized if "loss"?

[the remark this answers is not in the document]
Comment (sfujiwara) [reply optuna/integration/tensorflow.py]: Thanks, it's my mistake.
I fixed them as below:
```
It should be True if you use a metric to be maximize such as ``loss``.
==>
It should be :obj:`True` if you use a metric to be maximize such as ``accuracy``.
```

File: optuna/integration/tensorflow.py [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+    """
+    def __init__(self, trial, estimator, metric, is_higher_better, run_every_steps):
+        # type: (optuna.trial.Trial, tf.estimator.Estimator, str, bool, int) -> None
+
+        self.trial = trial
+        self.estimator = estimator
+        self.current_summary_step = -1
+        self.metric = metric
+        self.is_higher_better = is_higher_better
+        self._global_step_tensor = None
```
[the remark this answers is not in the document]
Comment (sfujiwara) [reply optuna/integration/tensorflow.py]: I fixed it deleting `_`.

File: optuna/integration/tensorflow.py (hunk 11 above)
Comment (sfujiwara) [reply optuna/integration/tensorflow.py]: I fixed them in L78 and L87.
You mentioned to L81, but is it L87?

Comment (sile) [reply optuna/integration/tensorflow.py]: Oops..., sorry, L87 is correct.

**Comment by sile** [comment]
@sfujiwara Thank you for fixing.  I'm sorry for the addition, but could you deal with https://github.com/pfnet/optuna/pull/292#issuecomment-452197504 ?

**Comment by sfujiwara** [comment]
@sile I added `TensorFlowPruningHook` to the integration list in `docs/source/tutorial/pruning.rst`. Please check it!
